PMM-6015 building PMM with a python script

nikita-b opened the pull request.

nikita-b committed:
PMM-6015 get rid of submodules

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-0ef9411
client docker - perconalab/pmm-client-fb:PR-1907-0ef9411
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-0ef9411.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-0ef9411&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-0ef9411.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7444/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/6862/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9246/

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-cce4a45
client docker - perconalab/pmm-client-fb:PR-1907-cce4a45
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-cce4a45.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-cce4a45&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-cce4a45.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7449/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/6867/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9254/

nikita-b committed:
fix

nikita-b committed:
PMM-8364 merge pmm-server and percona-images repos (#1892)

* PMM-8364 merge percona-images PMM content to pmm-server repo

nikita-b committed:
add requirements.txt

nikita-b committed:
fix bug with new release branch naming (#1911)

* fix bug with new release branch naming

nikita-b committed:
fix base name for branch (#1913)

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
remove submodules file

nikita-b committed:
ifx

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-e8fc610
client docker - perconalab/pmm-client-fb:PR-1907-e8fc610
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-e8fc610.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-e8fc610&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-e8fc610.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7512/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/6939/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9350/

nikita-b committed:
fix tests

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-68c2e05
client docker - perconalab/pmm-client-fb:PR-1907-68c2e05
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-68c2e05.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-68c2e05&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-68c2e05.tar.gz

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-f1c7b7f
client docker - perconalab/pmm-client-fb:PR-1907-f1c7b7f
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-f1c7b7f.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-f1c7b7f&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-f1c7b7f.tar.gz

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-7f9704b
client docker - perconalab/pmm-client-fb:PR-1907-7f9704b
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-7f9704b.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-7f9704b&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-7f9704b.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7521/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/6947/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9365/

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7523/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/6949/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9367/

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7524/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/6950/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9368/

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-da8561f
client docker - perconalab/pmm-client-fb:PR-1907-da8561f
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-da8561f.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-da8561f&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-da8561f.tar.gz

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-14758aa
client docker - perconalab/pmm-client-fb:PR-1907-14758aa
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-14758aa.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-14758aa&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-14758aa.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7525/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/6951/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9372/

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7526/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/6952/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9373/

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-f06bb72
client docker - perconalab/pmm-client-fb:PR-1907-f06bb72
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-f06bb72.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-f06bb72&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-f06bb72.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7538/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/6964/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9387/

nikita-b committed:
fix

nikita-b committed:
remove even more

nikita-b committed:
fix

nikita-b committed:
Merge branch 'PMM-2.0' of github.com:Percona-Lab/pmm-submodules into PMM-6015-reduce-build-time

nikita-b committed:
remove submodules again

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-e4e5d37
client docker - perconalab/pmm-client-fb:PR-1907-e4e5d37
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-e4e5d37.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-e4e5d37&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-e4e5d37.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7594/

ademidoff commented on .git-deps.yml:
```diff
@@ -0,0 +1,128 @@
+deps:
```
It would be great to rename this file to `ci.yml`, and `git-deps.py`  to `ci.py` so they clearly tell the user what they are used for. What do you think?

ademidoff commented on .git-deps.yml:
```diff
@@ -0,0 +1,128 @@
+  url: https://github.com/percona/pmm
+  component: common
+# CLIENT
+- name: pmm-admin
```
Could we put an empty line before each new `-name` section to make it more readable?

ademidoff commented on git-deps.py:
```diff
@@ -0,0 +1,127 @@
+
+DEFAULT_BRANCH = 'main' # we can rewrite it in config
+CONFIG_NAME = '.gitmodules-new'
+YAML_CONFIG = '.git-deps.yml'
```
```suggestion
YAML_CONFIG = 'ci.yml'
```

ademidoff commented on Makefile:
```diff
@@ -1,4 +1,4 @@
-.PHONY: all submodules server client clean purge test help default fb
+.PHONY: all server client prepare build clean purge help default
```
I see `purge` is gone, and `prepare` and `build` are not there... yet :) 
Maybe they exist in your local version, just need to push upstream?

ademidoff reviewed (changes requested):
Very powerful stuff, thanks!

ademidoff commented:
@nikita-b Just a question: will there be updates to relevant client & server Jenkins pipelines? If yes, maybe we could link them here?

nikita-b committed:
return submodules

nikita-b committed:
new

nikita-b committed:
Create fuature build

nikita-b committed:
Create fuature build

nikita-b committed:
Create fuature build

nikita-b committed:
Create fuature build

nikita-b committed:
Create fuature build

nikita-b committed:
Create fuature build

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
Merge branch 'PMM-2.0' of github.com:Percona-Lab/pmm-submodules into PMM-6015-reduce-build-time

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
add commit

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-7d5c615
client docker - perconalab/pmm-client-fb:PR-1907-7d5c615
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-7d5c615.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-7d5c615&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-7d5c615.tar.gz

nikita-b committed:
remove

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-de7548b
client docker - perconalab/pmm-client-fb:PR-1907-de7548b
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-de7548b.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-de7548b&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-de7548b.tar.gz

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-9cf8963
client docker - perconalab/pmm-client-fb:PR-1907-9cf8963
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-9cf8963.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-9cf8963&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-9cf8963.tar.gz

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-d0e0725
client docker - perconalab/pmm-client-fb:PR-1907-d0e0725
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-d0e0725.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-d0e0725&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-d0e0725.tar.gz

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9600/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9601/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9602/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9603/

ademidoff commented on Makefile:
```diff
@@ -1,4 +1,12 @@
-.PHONY: all submodules server client clean purge test help default fb
+.PHONY: all deps migrate server client prepare build clean purge fb help default
```
`make submodules` should still be possible, right?

ademidoff replied:
hm, there is no `prepare` in the code

BupycHuk replied:
+1, probably prepare is create, but I think prepare has more sense here.

ademidoff commented on README.md:
```diff
@@ -1 +1,33 @@
\ No newline at end of file
+# PMM Submodules
+
+You can create feature build for PMM in this repo.
```
```suggestion
This repository serves the purpose of creating and updating the feature build for PMM Server and PMM Client.
```

ademidoff commented on README.md:
```diff
@@ -1 +1,33 @@
+
+You can create feature build for PMM in this repo.
+
+## How to create feature build
```
```suggestion
## How to create a feature build
```

ademidoff commented on README.md:
```diff
@@ -1 +1,33 @@
+
+## How to create feature build
+
+if you use the same branch name in all repos then you can run:
```
```suggestion
If you use the same branch name in all repos then you can run:
```

ademidoff commented on README.md:
```diff
@@ -1 +1,33 @@
+Branches with "you branch name" will be used for all repos or default branch if the branch with this name isn't found in repo.
+
+
+## Q&A
```
```suggestion
## FAQ
```

ademidoff commented on README.md:
````diff
@@ -1 +1,33 @@
+```
+
+### What's global repo in ci.yml?
+It's name that script will try to find in repo instead default branch.
````
```suggestion
It's a branch name that this script will try to find in a repo instead of the default branch (usually called `main` or `PMM-2.0`).
```

ademidoff commented on ci-default.yml:
```diff
@@ -0,0 +1,118 @@
+  path: sources/pmm-admin/src/github.com/percona/pmm-admin
+  url: https://github.com/percona/pmm-admin.git
+  component: client
+- name: pmm-agent
```
Can we put a space line between two neighboring `-name` sections for better readability?

BupycHuk replied:
+1

ademidoff commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+
+logging.basicConfig(stream=sys.stdout, format='[%(levelname)s] %(asctime)s: %(message)s', level=logging.INFO)
+
+DEFAULT_BRANCH = 'main' # we can rewrite it in config
```
rewrite -> redefine maybe?

ademidoff replied:
Is it used in the code? I can't see where )

nikita-b committed:
fix

BupycHuk reviewed (changes requested):
One more argument to keep commit hashes in the repository is that community developers and we should be able to build docker container locally and result should be the same with how it will be built in jenkins. Uncontrollable changes is not traceable and leads to problems in checking history of builds.

ademidoff commented:
> One more argument to keep commit hashes in the repository is that community developers and we should be able to build docker container locally and result should be the same with how it will be built in jenkins. Uncontrollable changes is not traceable and leads to problems in checking history of builds.

Despite the title, this PR doesn't remove submodules. So one can continue building PMM as before.

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+        sys.exit(0)
+
+def switch_or_create_branch(path, branch):
+    # it's a small hack for migration from submodules
```
what this hack does and what problem did we have?

BupycHuk replied:
let's add more detailed info

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+        logging.info(f'Need to create PR now: https://github.com/Percona-Lab/pmm-submodules/compare/{branch_name}?expand=1')
+
+    def get_deps(self, single_branch=False):
+        with open(GIT_SOURCES_FILE, 'w+') as f:
+            f.truncate()
+
+        with open(GIT_SOURCES_FILE, 'a') as f:
```
let's just open with `w`, so it truncates file by itself.

nikita-b replied:
I used `append` mode here because if the script fails we'll have the last commits in `git-sources`. It's useful for debugging.

BupycHuk replied:
2 lines above you truncate the fiile, so append doesn't help. You will have empty file anyway.

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+            url = dep['url']
+            g = git.cmd.Git()
+            # TODO maybe it'll be faster to use local data
+            output = g.ls_remote("--heads", url, self.global_branch_name)
```
is output just a string or array? if string then check `if self.global_branch_name in output` will work for cases like
`PMM-somevalue` if we have remote branch `PMM-somevalue-extra-suffix` or `prefix-PMM-somevalue`
I suggest to use Github API client to request if branch exists.

nikita-b replied:
I changed this part to Github API

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+    def set_global_branches(self):
+        for dep in self.config['deps']:
+            url = dep['url']
+            g = git.cmd.Git()
```
let's use `git_cmd` name instead of just one letter for better readability.

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/env python3
+from pathlib import Path
```
Please add readme at the beginning explaining what does this tool do.

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+        if self.custom_config is not None:
+            # first we want to find global branch
+            for conf in self.custom_config['deps']:
+                if conf['name'] == 'global':
```
isn't it better to use global on `ci.yml` creation script to see what exact repositories are affected.

BupycHuk replied:
It makes debugging and reviewing even harder.

nikita-b replied:
If we'll add a new branch then we'll need to make a new commit.

I'll add information about the branches that were used.

BupycHuk replied:
Yes, and that should be transparent, we shouldn't go to jenkins to see new changes or try to see differences between two comments in PR. We have git, so let's use proper tools instead of making hacks. 
What's problem with having new commits?

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+                        dep['branch'] = conf['branch']
+                        break
+                else:
+                    logging.error(f'Can"t find {conf["name"]} repo. We have list of repos in ci-default.yml')
```
```suggestion
                    logging.error(f'Can"t find {conf["name"]} repo from ci.yml in the list of repos in ci-default.yml')
```

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+        self.custom_config = {}
+        self.read_custom_config()
+        self.read_config()
+        self.deps = self.config['deps']
```
is it required?

BupycHuk replied:
we use it only in one place, so it can be called via `self.config["deps"]`

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+            yaml.dump(config, f, sort_keys=False)
+
+
+    def read_config(self):
```
let's make this read_config return config value and assign self.config in constructor instead of updating self.config inside. it will be easier to read.

BupycHuk replied:
the same for `read_custom_config`

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+                logging.info(f'Use branch {self.global_branch_name} for {dep["name"]}')
+                dep['branch'] = self.global_branch_name
+
+    def create_fb(self, branch_name):
```
`create_fb` doesn't create FB actually, right? Maybe it's better to rename it to `prepare_deps`?

nikita-b replied:
It creates a branch for FB. We only need to create PR but I hope we'll implement it too.

BupycHuk replied:
`create_fb_branch` then. Code should be self-explaining, do not mislead the ones who will read the code.
Creating branch for FB is a part of actual creating FB process which includes:
1. Create FB branch
2. Create PR
3. Build FB

nikita-b replied:
Renamed.

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+
+        self.write_custom_config(self.custom_config)
+        repo.git.add(all=True)
+        repo.index.commit('Create fuature build')
```
```suggestion
        repo.index.commit('Create feature build')
```

BupycHuk replied:
let's add branch name into commit message, so it's visible in Jira.

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+            self.custom_config = {'deps': [global_branch,]}
+
+        self.write_custom_config(self.custom_config)
+        repo.git.add(all=True)
```
let's add exact file instead of all

BupycHuk commented on README.md:
````diff
@@ -1 +1,33 @@
+git push
+```
+
+### What's global repo in ci.yml?
````
having `global` on get deps staging adds more risks then benefits, I'm pretty sure that everybody will use it and it will be harder to figure out from submodules which repos were affected. So it brings the case when different team members might review only PRs requested from them without knowing about other PRs.

nikita-b replied:
We can send comment to PR with the list of branches for example. Or edit the first comment in PR?

BupycHuk replied:
It's a hack, let's use git to trace changes instead of trying to hide them. If you want to make it easier for developers to rebuild PR in case of new repo added then it's better to create new jenkins job or github action to rewrite ci.yml and commit.

nikita-b replied:
Changed and now global work only on the first run.

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+                    if single_branch:
+                        check_call(f'git clone --depth 1 --single-branch --branch {target_branch} {target_url} {path}'.split())
+                    else:
+                        check_call(f'git clone --depth 1 --no-single-branch {target_url} {path}')
```
what case we might want to clone all branches in?

nikita-b replied:
I was just comparing the difference in speed. Removed no-single-branch version

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+                    else:
+                        check_call(f'git clone --depth 1 --no-single-branch {target_url} {path}')
+                else:
+                    print('Path for {} already exist'.format(dep["name"]))
```
```suggestion
                    print('Files in the path for {} is already exist'.format(dep["name"]))
```

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+        with open(GIT_SOURCES_FILE, 'a') as f:
+            for dep in self.deps:
+                path = os.path.join(self.rootdir, dep["path"])
+                if not os.path.exists(os.path.join(self.rootdir, path)):
```
if you will keep sources directories with submodules, then directory with this path always will be exist.
I suggest to remove all directories in `sources` directory first and then clone or go to `submodule` directory and pull required branch.

nikita-b replied:
It works well with the submodules directory. We need this part only if will remove submodules.

BupycHuk replied:
Oh, I see, you switch branch inside submodule, right?

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,200 @@
+        self.write_custom_config(self.custom_config)
+        repo.git.add(all=True)
+        repo.index.commit('Create fuature build')
+        origin = repo.remote(name='origin')
```
this command doesn't push changes to remote, so you should push first before creating PR.

nikita-b replied:
Yep, I just deleted it when tested.

nikita-b committed:
fisrt part of fixes

BupycHuk commented:
@atymchuk so if it doesn't get rid of submodules, then PR title should be renamed to not mislead reviewers.

BupycHuk commented on Makefile:
```diff
@@ -1,4 +1,12 @@
-.PHONY: all submodules server client clean purge test help default fb
+.PHONY: all submodules deps prepare server client build clean purge fb help default
+
+# Ugly hack... 2021 year and we still use make...God forgive us
```
this comment doesn't add more info

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,210 @@
+                dep['branch'] = self.global_branch_name
+
+    def create_fb(self, branch_name):
+        import git
```
we have import git in the top

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,210 @@
+        import git
+        repo = git.Repo('.')
+
+        git = repo.git
```
having `git` in imports and having `git` as variable confuses people.

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,210 @@
+            yaml.dump(self.submodules, f, sort_keys=False)
+        sys.exit(0)
+
+def switch_or_create_branch(path, branch):
```
we shouldn't create branch in case of some mistakes in branch name, so it's better to show error instead of silently creating new branch.

BupycHuk commented on Jenkinsfile:
```diff
@@ -57,39 +57,57 @@ pipeline {
-                    echo $pmm_ui_tests_branch > pmmUITestBranch
-                    echo $pmm_ui_tests_commit_sha > pmmUITestsCommitSha
-                    cd $curdir
+                    if [ -s ci.yml ]
```
by the way in slack you said that build process for dev-latest will not be affected, but you included ci.yml into this PR which means that we change default behavior.

nikita-b replied:
We don't change default behavior `-s` means `file isn't empty`:
https://tldp.org/LDP/abs/html/fto.html

BupycHuk replied:
it's good to add comment there.

nikita-b committed:
add github api

nikita-b committed:
rename create fb

nikita-b committed:
fix

nikita-b committed:
Merge branch 'PMM-2.0' of github.com:Percona-Lab/pmm-submodules into PMM-6015-reduce-build-time

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-d197d4f
client docker - perconalab/pmm-client-fb:PR-1907-d197d4f
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-d197d4f.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-d197d4f&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-d197d4f.tar.gz

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9687/

nikita-b committed:
fix

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-cdf873d
client docker - perconalab/pmm-client-fb:PR-1907-cdf873d
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-cdf873d.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-cdf873d&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-cdf873d.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7758/

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-743f580
client docker - perconalab/pmm-client-fb:PR-1907-743f580
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-743f580.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-743f580&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-743f580.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7771/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/9715/

nikita-b committed:
fix

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-99d0dae
client docker - perconalab/pmm-client-fb:PR-1907-99d0dae
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-99d0dae.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-99d0dae&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-99d0dae.tar.gz

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-c98589e
client docker - perconalab/pmm-client-fb:PR-1907-c98589e
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-c98589e.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-c98589e&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-c98589e.tar.gz

nikita-b committed:
Merge branch 'PMM-2.0' of github.com:Percona-Lab/pmm-submodules into PMM-6015-reduce-build-time

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-ba12d6b
client docker - perconalab/pmm-client-fb:PR-1907-ba12d6b
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-ba12d6b.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-ba12d6b&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-ba12d6b.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7773/

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7774/

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7775/

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix requestments

nikita-b committed:
fix

nikita-b committed:
fix

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-7a1e774
client docker - perconalab/pmm-client-fb:PR-1907-7a1e774
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-7a1e774.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-7a1e774&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-7a1e774.tar.gz

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-da736a9
client docker - perconalab/pmm-client-fb:PR-1907-da736a9
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-da736a9.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-da736a9&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-da736a9.tar.gz

nikita-b committed:
fix

nikita-b committed:
fix

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7779/

JNKPercona commented:
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/7211/

nikita-b committed:
ifx

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7780/

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-095642b
client docker - perconalab/pmm-client-fb:PR-1907-095642b
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-095642b.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-095642b&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-095642b.tar.gz

nikita-b committed:
fix

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-441e14e
client docker - perconalab/pmm-client-fb:PR-1907-441e14e
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-441e14e.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-441e14e&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-441e14e.tar.gz

nikita-b committed:
fix

nikita-b commented:
@BupycHuk @atymchuk I updated PR.
1. Changed behavior for `global` option.
2. Github has a limit for API requests (50 requests in hours) without auth (It's enough only for 2 PR). I added token support and if you use token then you have a bigger limit (1000). 
3.  If you use token, then the script also creates PR.
4. Also, I fixed all discussion from @BupycHuk

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7782/

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-89dc851
client docker - perconalab/pmm-client-fb:PR-1907-89dc851
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-89dc851.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-89dc851&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-89dc851.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7784/

nikita-b committed:
fix tests

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7785/

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-537e2e6
client docker - perconalab/pmm-client-fb:PR-1907-537e2e6
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-537e2e6.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-537e2e6&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-537e2e6.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7786/

Commit:
chore: update README.md

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-1907-2858dc5
client docker - perconalab/pmm-client-fb:PR-1907-2858dc5
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-2858dc5.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-1907-2858dc5&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-1907-2858dc5.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/7791/

denisok commented on ci.py:
```diff
@@ -0,0 +1,233 @@
+            for conf in self.custom_config['deps']:
+                for dep in self.config['deps']:
+                    if dep['name'] == conf['name']:
+                        # TODO add support for other fields
```
right, I think we would need `url` for the external contributors to point to their forks.

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,233 @@
+            github_api = Github(GITHUB_TOKEN)
+            repo = github_api.get_repo(repo_path)
+
+            for branch in repo.get_branches():
```
repo has `get_branch`, why don't use it?

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,233 @@
+            repo = github_api.get_repo('Percona-Lab/pmm-submodules')
+            pr = repo.get_pulls(base='PMM-2.0', head=f'Percona-Lab:{branch_name}')
+            # TODO we can use totalCount here: https://github.com/PyGithub/PyGithub/blob/babcbcd04fd5605634855f621b8558afc5cbc515/github/PaginatedList.py#L102
+            # but it works pretty strange. It reterned count ALL PR from repo without filters
```
```suggestion
            # but it works pretty strange. It returned count ALL PR from repo without filters
```

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,233 @@
+            # TODO we can use totalCount here: https://github.com/PyGithub/PyGithub/blob/babcbcd04fd5605634855f621b8558afc5cbc515/github/PaginatedList.py#L102
+            # but it works pretty strange. It reterned count ALL PR from repo without filters
+            hasPR = False
+            for i in pr:
+                hasPR = True
+                break
+            if not hasPR:
```
it could be `for else`

BupycHuk commented on ci.py:
```diff
@@ -0,0 +1,233 @@
+        origin = repo.remote(name='origin')
+        origin.push()
+        logging.info('Last ci.yml was pushed')
+        if GITHUB_TOKEN:
```
👍🏼
